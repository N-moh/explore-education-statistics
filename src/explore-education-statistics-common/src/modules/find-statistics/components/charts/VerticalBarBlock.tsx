import {
  ChartDataB,
  ChartDefinition,
  conditionallyAdd,
  createDataForAxis,
  getKeysForChart,
  mapNameToNameLabel,
  populateDefaultChartProps,
  StackedBarProps,
} from '@common/modules/find-statistics/components/charts/ChartFunctions';
import React, { Component } from 'react';
import {
  Bar,
  BarChart,
  CartesianGrid,
  Legend,
  ReferenceLine,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';

import classnames from 'classnames';

import './charts.scss';

export default class VerticalBarBlock extends Component<StackedBarProps> {
  public static definition: ChartDefinition = {
    type: 'verticalbar',
    name: 'Vertical bar',

    capabilities: {
      dataSymbols: false,
      stackable: true,
      lineStyle: false,
      gridLines: true,
    },

    data: [
      {
        type: 'bar',
        title: 'Bar',
        entryCount: 1,
        targetAxis: 'xaxis',
      },
    ],

    axes: [
      {
        id: 'major',
        title: 'X Axis',
        type: 'major',
        defaultDataType: 'timePeriods',
      },
      {
        id: 'minor',
        title: 'Y Axis',
        type: 'minor',
      },
    ],
  };

  public render() {
    const {
      data,
      meta,
      height,
      width,
      labels,
      axes,
      stacked,
      legend,
      legendHeight,
    } = this.props;

    if (axes.major === undefined || data === undefined || meta === undefined)
      return <div>Unable to render chart</div>;

    const chartData: ChartDataB[] = createDataForAxis(
      axes.major,
      data.result,
      meta,
    ).map(mapNameToNameLabel(labels, meta.timePeriods, meta.locations));

    const keysForChart = getKeysForChart(chartData);

    return (
      <ResponsiveContainer width={width || 900} height={height || 300}>
        <BarChart
          data={chartData}
          className={classnames({ 'legend-bottom': legend === 'bottom' })}
          margin={{
            left: 30,
            top: legend === 'top' ? 10 : 0,
          }}
        >
          <CartesianGrid
            strokeDasharray="3 3"
            vertical={axes.minor && axes.minor.showGrid !== false}
            horizontal={axes.major && axes.major.showGrid !== false}
          />

          {axes.minor && (
            <YAxis
              type="number"
              hide={axes.minor.visible === false}
              label={{
                angle: -90,
                offset: 0,
                position: 'left',
                value: '',
              }}
              scale="auto"
              width={conditionallyAdd(axes.minor && axes.minor.size)}
              interval={
                axes.minor && !axes.minor.visible
                  ? 'preserveStartEnd'
                  : undefined
              }
            />
          )}

          {axes.major && (
            <XAxis
              type="category"
              dataKey="name"
              hide={axes.major.visible === false}
              label={{
                offset: 5,
                position: 'bottom',
                value: '',
              }}
              scale="auto"
              padding={{ left: 20, right: 20 }}
              height={conditionallyAdd(
                axes.major && axes.major.size,
                legend === 'bottom' ? 50 : undefined,
              )}
              tickMargin={10}
            />
          )}

          <Tooltip />
          {(legend === 'top' || legend === 'bottom') && (
            <Legend
              verticalAlign={legend}
              height={+(legendHeight || '50')}
              margin={{ top: 5, bottom: 5 }}
            />
          )}

          {Array.from(keysForChart).map(name => (
            <Bar
              key={name}
              {...populateDefaultChartProps(name, labels[name])}
              stackId={stacked ? 'a' : undefined}
            />
          ))}

          {axes.major &&
            axes.major.referenceLines &&
            axes.major.referenceLines.map(referenceLine => (
              <ReferenceLine
                key={`${referenceLine.position}_${referenceLine.label}`}
                x={referenceLine.position}
                label={referenceLine.label}
              />
            ))}

          {axes.minor &&
            axes.minor.referenceLines &&
            axes.minor.referenceLines.map(referenceLine => (
              <ReferenceLine
                key={`${referenceLine.position}_${referenceLine.label}`}
                y={referenceLine.position}
                label={referenceLine.label}
              />
            ))}
        </BarChart>
      </ResponsiveContainer>
    );
  }
}
