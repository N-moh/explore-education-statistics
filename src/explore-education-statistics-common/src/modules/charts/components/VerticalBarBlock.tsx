import '@common/modules/charts/components/charts.scss';
import {
  ChartDefinition,
  StackedBarProps,
} from '@common/modules/charts/types/chart';
import {
  ChartData,
  conditionallyAdd,
  createSortedAndMappedDataForAxis,
  generateMajorAxis,
  generateMinorAxis,
  getKeysForChart,
  populateDefaultChartProps,
} from '@common/modules/charts/util/chartUtils';
import React from 'react';
import {
  Bar,
  BarChart,
  CartesianGrid,
  Legend,
  ReferenceLine,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';

export type VerticalBarProps = StackedBarProps;

const VerticalBarBlock = ({
  data,
  meta,
  height,
  width,
  labels,
  axes,
  stacked,
  legend,
  renderLegend,
}: VerticalBarProps) => {
  if (
    axes === undefined ||
    axes.major === undefined ||
    axes.minor === undefined ||
    data === undefined ||
    meta === undefined
  )
    return <div>Unable to render chart, chart incorrectly configured</div>;

  const chartData: ChartData[] = createSortedAndMappedDataForAxis(
    axes.major,
    data.result,
    meta,
    labels,
  );

  const keysForChart = getKeysForChart(chartData);

  const minorDomainTicks = generateMinorAxis(chartData, axes.minor);
  const majorDomainTicks = generateMajorAxis(chartData, axes.major);

  return (
    <ResponsiveContainer width={width || '100%'} height={height || 300}>
      <BarChart
        data={chartData}
        margin={{
          left: 30,
        }}
      >
        <CartesianGrid
          strokeDasharray="3 3"
          vertical={axes.minor?.showGrid !== false}
          horizontal={axes.major.showGrid !== false}
        />

        <YAxis
          type="number"
          dataKey="value"
          hide={axes.minor?.visible === false}
          unit={axes.minor?.unit}
          scale="auto"
          {...minorDomainTicks}
          width={conditionallyAdd(axes.minor?.size)}
        />

        <XAxis
          type="category"
          dataKey="name"
          hide={axes.major.visible === false}
          unit={axes.major.unit}
          scale="auto"
          {...majorDomainTicks}
          padding={{ left: 20, right: 20 }}
          tickMargin={10}
        />

        <Tooltip />

        {legend && legend !== 'none' && (
          <Legend content={renderLegend} align="left" layout="vertical" />
        )}

        {keysForChart.map(name => (
          <Bar
            key={name}
            {...populateDefaultChartProps(name, labels[name])}
            stackId={stacked ? 'a' : undefined}
          />
        ))}

        {axes.major.referenceLines?.map(referenceLine => (
          <ReferenceLine
            key={`${referenceLine.position}_${referenceLine.label}`}
            x={referenceLine.position}
            label={referenceLine.label}
          />
        ))}

        {axes.minor?.referenceLines?.map(referenceLine => (
          <ReferenceLine
            key={`${referenceLine.position}_${referenceLine.label}`}
            y={referenceLine.position}
            label={referenceLine.label}
          />
        ))}
      </BarChart>
    </ResponsiveContainer>
  );
};

const definition: ChartDefinition = {
  type: 'verticalbar',
  name: 'Vertical bar',

  capabilities: {
    dataSymbols: false,
    stackable: true,
    lineStyle: false,
    gridLines: true,
    canSize: true,
    fixedAxisGroupBy: false,
    hasAxes: true,
    hasReferenceLines: true,
    hasLegend: true,
  },

  data: [
    {
      type: 'bar',
      title: 'Bar',
      entryCount: 1,
      targetAxis: 'xaxis',
    },
  ],

  axes: [
    {
      id: 'major',
      title: 'X Axis',
      type: 'major',
      defaultDataType: 'timePeriod',
    },
    {
      id: 'minor',
      title: 'Y Axis',
      type: 'minor',
    },
  ],

  requiresGeoJson: false,
};

VerticalBarBlock.definition = definition;

export default VerticalBarBlock;
