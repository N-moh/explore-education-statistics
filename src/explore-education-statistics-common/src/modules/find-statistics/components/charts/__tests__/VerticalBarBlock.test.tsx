import React from 'react';

import { render } from 'react-testing-library';
import VerticalBarBlock from '../VerticalBarBlock';

import testData from './__data__/testBlockData';

jest.mock('recharts/lib/util/LogUtils');

describe('VerticalBarBlock', () => {
  test('renders with correct output', () => {
    const { container } = render(
      <VerticalBarBlock {...testData.AbstractChartProps} />,
    );

    /*
    expect(
      container.querySelector('.xAxis text.recharts-label tspan'),
    ).toHaveTextContent('test x axis');
    expect(
      container.querySelector('.yAxis text.recharts-label tspan'),
    ).toHaveTextContent('test y axis');

     */

    expect(
      container.querySelector(
        '.xAxis .recharts-cartesian-axis-tick:nth-child(1) text tspan',
      ),
    ).toHaveTextContent('2014/15');
    expect(
      container.querySelector(
        '.xAxis .recharts-cartesian-axis-tick:nth-child(2) text tspan',
      ),
    ).toHaveTextContent('2015/16');

    expect(Array.from(container.querySelectorAll('.recharts-bar')).length).toBe(
      3,
    );

    expect(container).toMatchSnapshot();
  });
});
