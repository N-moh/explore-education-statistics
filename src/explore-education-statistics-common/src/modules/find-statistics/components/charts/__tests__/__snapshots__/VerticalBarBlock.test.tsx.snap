// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalBarBlock renders with correct output 1`] = `
<div>
  <div
    class="recharts-responsive-container"
    style="width: 900px; height: 300px;"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 900px; height: 300px;"
    >
      <svg
        class="recharts-surface"
        height="300"
        version="1.1"
        viewBox="0 0 900 300"
        width="900"
      >
        <defs>
          <clippath
            id="recharts4-clip"
          >
            <rect
              height="235"
              width="785"
              x="85"
              y="15"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="25"
            stroke="#666"
            width="785"
            x="85"
            x1="85"
            x2="870"
            y="250"
            y1="250"
            y2="250"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="25"
                stroke="#666"
                width="785"
                x="85"
                x1="281.25"
                x2="281.25"
                y="250"
                y1="256"
                y2="250"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                font-size="12"
                height="25"
                stroke="none"
                text-anchor="middle"
                width="785"
                x="281.25"
                y="258"
              >
                <tspan
                  dy="0.71em"
                  x="281.25"
                >
                  2014/15
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="25"
                stroke="#666"
                width="785"
                x="85"
                x1="673.75"
                x2="673.75"
                y="250"
                y1="256"
                y2="250"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                font-size="12"
                height="25"
                stroke="none"
                text-anchor="middle"
                width="785"
                x="673.75"
                y="258"
              >
                <tspan
                  dy="0.71em"
                  x="673.75"
                >
                  2015/16
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-cartesian-grid"
        >
          <g
            class="recharts-cartesian-grid-horizontal"
          >
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="85"
              x2="870"
              y="15"
              y1="250"
              y2="250"
            />
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="85"
              x2="870"
              y="15"
              y1="191.25"
              y2="191.25"
            />
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="85"
              x2="870"
              y="15"
              y1="132.5"
              y2="132.5"
            />
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="85"
              x2="870"
              y="15"
              y1="73.75"
              y2="73.75"
            />
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="85"
              x2="870"
              y="15"
              y1="15"
              y2="15"
            />
          </g>
          <g
            class="recharts-cartesian-grid-vertical"
          >
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="281.25"
              x2="281.25"
              y="15"
              y1="15"
              y2="250"
            />
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="673.75"
              x2="673.75"
              y="15"
              y1="15"
              y2="250"
            />
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="85"
              x2="85"
              y="15"
              y1="15"
              y2="250"
            />
            <line
              fill="none"
              height="235"
              stroke="#ccc"
              width="785"
              x="85"
              x1="870"
              x2="870"
              y="15"
              y1="15"
              y2="250"
            />
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="235"
            stroke="#666"
            width="25"
            x="60"
            x1="85"
            x2="85"
            y="15"
            y1="15"
            y2="250"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="235"
                stroke="#666"
                width="25"
                x="60"
                x1="79"
                x2="85"
                y="15"
                y1="250"
                y2="250"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="235"
                stroke="none"
                text-anchor="end"
                width="25"
                x="77"
                y="250"
              >
                <tspan
                  dy="0.355em"
                  x="77"
                >
                  -3
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="235"
                stroke="#666"
                width="25"
                x="60"
                x1="79"
                x2="85"
                y="15"
                y1="191.25"
                y2="191.25"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="235"
                stroke="none"
                text-anchor="end"
                width="25"
                x="77"
                y="191.25"
              >
                <tspan
                  dy="0.355em"
                  x="77"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="235"
                stroke="#666"
                width="25"
                x="60"
                x1="79"
                x2="85"
                y="15"
                y1="132.5"
                y2="132.5"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="235"
                stroke="none"
                text-anchor="end"
                width="25"
                x="77"
                y="132.5"
              >
                <tspan
                  dy="0.355em"
                  x="77"
                >
                  3
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="235"
                stroke="#666"
                width="25"
                x="60"
                x1="79"
                x2="85"
                y="15"
                y1="73.75"
                y2="73.75"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="235"
                stroke="none"
                text-anchor="end"
                width="25"
                x="77"
                y="73.75"
              >
                <tspan
                  dy="0.355em"
                  x="77"
                >
                  6
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <line
                class="recharts-cartesian-axis-tick-line"
                fill="none"
                height="235"
                stroke="#666"
                width="25"
                x="60"
                x1="79"
                x2="85"
                y="15"
                y1="15"
                y2="15"
              />
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="235"
                stroke="none"
                text-anchor="end"
                width="25"
                x="77"
                y="15"
              >
                <tspan
                  dy="0.355em"
                  x="77"
                >
                  9
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-bar"
        >
          <g
            class="recharts-layer recharts-bar-rectangles"
          >
            <g
              class="recharts-layer"
            >
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-bar"
        >
          <g
            class="recharts-layer recharts-bar-rectangles"
          >
            <g
              class="recharts-layer"
            >
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-bar"
        >
          <g
            class="recharts-layer recharts-bar-rectangles"
          >
            <g
              class="recharts-layer"
            >
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
              <g
                class="recharts-layer recharts-bar-rectangle"
              />
            </g>
          </g>
        </g>
      </svg>
      <div
        class="recharts-legend-wrapper"
        style="position: absolute; width: 810px; height: auto; left: 60px; bottom: 25px;"
      >
        <ul
          class="recharts-default-legend"
          style="padding: 0px; margin: 0px; text-align: center;"
        >
          <li
            class="recharts-legend-item legend-item-0"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              version="1.1"
              viewBox="0 0 32 32"
              width="14"
            >
              <path
                class="recharts-legend-icon"
                d="M0,4h32v24h-32z"
                fill="#4763a5"
                stroke="none"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
            >
              Unauthorised absence rate
            </span>
          </li>
          <li
            class="recharts-legend-item legend-item-1"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              version="1.1"
              viewBox="0 0 32 32"
              width="14"
            >
              <path
                class="recharts-legend-icon"
                d="M0,4h32v24h-32z"
                fill="#f5a450"
                stroke="none"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
            >
              Overall absence rate
            </span>
          </li>
          <li
            class="recharts-legend-item legend-item-2"
            style="display: inline-block; margin-right: 10px;"
          >
            <svg
              class="recharts-surface"
              height="14"
              style="display: inline-block; vertical-align: middle; margin-right: 4px;"
              version="1.1"
              viewBox="0 0 32 32"
              width="14"
            >
              <path
                class="recharts-legend-icon"
                d="M0,4h32v24h-32z"
                fill="#005ea5"
                stroke="none"
              />
            </svg>
            <span
              class="recharts-legend-item-text"
            >
              Authorised absence rate
            </span>
          </li>
        </ul>
      </div>
      <div
        class="recharts-tooltip-wrapper"
        style="pointer-events: none; visibility: hidden; position: absolute; top: 0px; transform: translate(undefinedpx, undefinedpx);"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
    <div
      style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
    />
  </div>
</div>
`;
